File: ui-automation-framework/pages/cart.page.ts
import type { Locator, Page } from '@playwright/test';
import { HeaderComponent } from '../components/header.component';
import { InventoryItemComponent } from '../components/inventory-item.component';

export class CartPage {
  readonly header: HeaderComponent;
  readonly items: Locator;
  readonly itemNames: Locator;
  readonly checkoutButton: Locator;
  readonly continueShoppingButton: Locator;

  constructor(readonly page: Page) {
    this.header = new HeaderComponent(page);
    this.items = page.getByTestId('inventory-item');
    this.itemNames = page.getByTestId('inventory-item-name');
    this.checkoutButton = page.getByTestId('checkout');
    this.continueShoppingButton = page.getByTestId('continue-shopping');
  }

  async goto(): Promise<void> {
    await this.page.goto('/cart.html');
  }

  item(productName: string): InventoryItemComponent {
    const root = this.items.filter({
      has: this.page.getByTestId('inventory-item-name').getByText(productName, { exact: true }),
    });
    return new InventoryItemComponent(root);
  }

  quantityOf(productName: string): Locator {
    return this.item(productName).root.getByTestId('item-quantity');
  }

  async removeItem(productName: string): Promise<void> {
    await this.item(productName).removeButton.click();
  }

  async startCheckout(): Promise<void> {
    await this.checkoutButton.click();
  }
}

File: ui-tests/src/test/java/saucedemo/pages/CartPage.java
package saucedemo.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import saucedemo.components.Header;
import saucedemo.components.ProductItem;

public class CartPage {
    private final Page page;
    public final Header header;

    public CartPage(Page page) {
        this.page = page;
        this.header = new Header(page);
    }

    public void open() {
        page.navigate("/cart.html");
    }

    public Locator items() {
        return page.locator("[data-test='inventory-item']");
    }

    public ProductItem item(String name) {
        for (int i = 0; i < items().count(); i++) {
            Locator one = items().nth(i);
            if (one.locator("[data-test='inventory-item-name']").innerText().equals(name)) {
                return new ProductItem(one);
            }
        }
        throw new RuntimeException("item not in cart: " + name);
    }

    public String quantityOf(String name) {
        Locator row = null;
        for (int i = 0; i < items().count(); i++) {
            Locator one = items().nth(i);
            if (one.locator("[data-test='inventory-item-name']").innerText().equals(name)) {
                row = one;
                break;
            }
        }
        if (row == null) {
            throw new RuntimeException("item not in cart: " + name);
        }
        return row.locator("[data-test='item-quantity']").innerText();
    }

    public void removeItem(String name) {
        item(name).removeFromCart();
    }

    public void continueShopping() {
        page.locator("[data-test='continue-shopping']").click();
    }

    public void startCheckout() {
        page.locator("[data-test='checkout']").click();
    }
}
